### PR TITLE
Getter/Setter functions for Firestore

### PR DESCRIPTION
**Pull request by kennethlien**
## Summary
- Created interfaces describing Tree, Comment, and Additional entries in the database.
- Added getter/setter functions for getting a single entry in Trees, Comments, or Additional with an ID from Firebase.
- Added function to get all Trees from Firebase.

## Test Plan
- Manual testing was done to confirm validity of results.

## Other
- Last object in `getAllTrees` occasionally has `undefined` as the key. This may be due to a Promise not resolving correctly.

### Next Steps
- Use in app to speed up development.
- 
### Relevant Links
https://www.notion.so/calblueprint/Querying-Firestore-029b6c2c3f5f4f639173d75622853234

### CODE REVIEW

File: src/database/firebase.tsx [hunk 1]
```diff
@@ -8,6 +8,7 @@ import {
   MESSAGING_SENDER_ID,
   APP_ID,
 } from '@env';
+import { Tree, Comment, Additional, Dictionary } from '../../types';
```
Comment (mylesdomingo) [inline src/database/firebase.tsx:11]: booo no relative imports 

`import {...} from '@types';`

File: src/database/firebase.tsx [hunk 2]
```diff
@@ -18,6 +19,149 @@ const config = {
...
+  return false;
+};
+
+/*
+ * getComment queries the `comments` table and returns a Comment if the ID is found and an empty entry otherwise.
+ */
+export const getComment = async function (id: string): Promise<Comment> {
+  await commentCollection
+    .doc(id)
+    .get()
```
Comment (mylesdomingo) [inline src/database/firebase.tsx:99]: wait I think you're not using async / await syntax correctly here. you should either wrap it in a try catch block or just use promises -- eg.

```
try { 
await treeCollection.doc(id).set({tree});
} catch (e) {
console.warn(e)
}
```

File: types.tsx [hunk 3]
```diff
@@ -35,3 +35,32 @@ export type RootTabScreenProps<Screen extends keyof RootTabParamList> =
...
+  longitude: number;
+};
+
+export type Planted = {
+  seconds: number;
+  nanoseconds: number;
+};
+
+export type Comment = {
+  [key: string]: string;
```
Comment (mylesdomingo) [inline types.tsx:57]: why dont comments have id's

File: src/database/firebase.tsx [hunk 4]
```diff
@@ -18,6 +19,149 @@ const config = {
...
+      console.log('Error getting document:', error);
+    });
+  return {} as Comment;
+};
+
+/*
+ * setComment creates/updates an entry in the `comments` table given a Comment.
+ */
+export const setComment = async function (
+  id: string,
```
Comment (mylesdomingo) [inline src/database/firebase.tsx:115]: shouldnt this just be 1 prop param eg comment that has an id (see below)

File: src/database/firebase.tsx [hunk 5]
```diff
@@ -18,6 +19,149 @@ const config = {
...
+  comment: Comment,
+): Promise<boolean> {
+  await commentCollection
+    .doc(id)
+    .set({ comment })
+    .then(() => {
+      return true;
```
Comment (mylesdomingo) [inline src/database/firebase.tsx:122]: do you need a boolean here for success? if the function doesn't throw you should consider success

File: src/database/firebase.tsx [hunk 6]
```diff
@@ -18,6 +19,149 @@ const config = {
...
+};
+
+/*
+ * getTree queries the `trees` table and returns a Tree if the ID is found and an empty entry otherwise.
+ */
+export const getTree = async function (id: string): Promise<Tree> {
+  await treeCollection
+    .doc(id)
+    .get()
+    .then(doc => {
```
Comment (mylesdomingo) [inline src/database/firebase.tsx:52]:
```
const entry = await treeCollection.doc(id).get() // assuming that .get() is the async function
return entry.data()
```

**Review by mylesdomingo — CHANGES_REQUESTED** [review]

File: src/database/firebase.tsx (hunk 6 above)
Comment (kennethlien) [reply src/database/firebase.tsx:52]: No error catching?

File: types.tsx [hunk 7]
```diff
@@ -35,3 +35,32 @@ export type RootTabScreenProps<Screen extends keyof RootTabParamList> =
...
+  planted: Planted | null;
+};
+
+export type Location = {
+  latitude: number;
+  longitude: number;
+}
+
+export type Planted = {
+  seconds: number;
```
[the remark this answers is not in the document]
Comment (kennethlien) [reply types.tsx:52]: Firestore timestamp is stored with seconds and nanoseconds.

File: src/database/firebase.tsx [hunk 8]
```diff
@@ -18,6 +19,149 @@ const config = {
...
+    .catch(error => {
+      console.log('Error getting document:', error);
+    });
+  return false;
+};
+
+/*
+ * getTree queries the `trees` table and returns a Tree if the ID is found and an empty entry otherwise.
+ */
+export const getTree = async function (id: string): Promise<Tree> {
```
Comment (mylesdomingo) [inline src/database/firebase.tsx:49]: these need to be arrow functions 

```
export const setAdditional = async (
  id: string,
  additional: Additional,
): Promise<boolean> => {
  await additionalCollection.doc(id).set({ additional });
  return false;
};

```

File: src/database/firebase.tsx [hunk 9]
```diff
@@ -18,6 +19,149 @@ const config = {
...
+    .catch(error => {
+      console.log('Error getting document:', error);
+    });
+  return { id: '' } as Tree;
+};
+
+/*
+ * getAllTrees returns a Dictionary containing all entries in the `trees` table.
+ */
+export const getAllTrees = async function (): Promise<Dictionary> {
```
Comment (wangannie) [inline src/database/firebase.tsx:64]: Why return a Dictionary? Seems more intuitive to have the default return an array of trees but maybe rename this to like `getTreesDictionary` or smth like that?

File: src/database/firebase.tsx (hunk 2 above)
Comment (kennethlien) [reply src/database/firebase.tsx]: Agreed

File: src/database/firebase.tsx (hunk 5 above)
Comment (kennethlien) [reply src/database/firebase.tsx]: made change

File: types.tsx [hunk 10]
```diff
@@ -12,7 +12,7 @@ import { NativeStackScreenProps } from '@react-navigation/native-stack';
 
 declare global {
   namespace ReactNavigation {
-    interface RootParamList extends RootStackParamList {}
+    interface RootParamList extends RootStackParamList { }
```
[the remark this answers is not in the document]
Comment (kennethlien) [reply types.tsx]: Yea, I had a different linter that ran after too...

File: src/database/firebase.tsx (hunk 9 above)
Comment (kennethlien) [reply src/database/firebase.tsx]: changed to array

File: src/database/firebase.tsx [hunk 11]
```diff
@@ -18,6 +19,149 @@ const config = {
...
+ * checkID validates that this ID exists in the `trees` table.
+ */
+export const checkID = async function (id: string): Promise<boolean> {
+  await treeCollection
+    .doc(id)
+    .get()
+    .then(doc => {
+      return doc.exists;
+    })
+    .catch(error => {
```
[the remark this answers is not in the document]
Comment (kennethlien) [reply src/database/firebase.tsx:39]: todo's added, all functions now throw error

**Review by mylesdomingo — CHANGES_REQUESTED** [review]

File: src/database/firebase.tsx [hunk 12]
```diff
@@ -18,6 +19,115 @@ const config = {
...
+const database = firebase.firestore();
+const treeCollection = database.collection('trees');
+const commentCollection = database.collection('comments');
+const additionalCollection = database.collection('additional');
+
+/*
+ * checkID validates that this ID exists in the `trees` table.
+ */
+export const checkID = async function (id: string): Promise<boolean> {
+  await treeCollection
```
Comment (mylesdomingo) [inline src/database/firebase.tsx:33]: I think you missed one

**Review by mylesdomingo — CHANGES_REQUESTED** [review]

File: src/database/firebase.tsx [hunk 13]
```diff
@@ -18,6 +19,115 @@ const config = {
...
+    .doc(id)
+    .get()
+    .then(doc => {
+      return doc.exists;
+    })
+    .catch(e => {
+      console.log('Error getting document:', e);
```
Comment (mylesdomingo) [inline src/database/firebase.tsx:40]: use console.warn(..) here for linting and also so its classified as an error / warning in the console

**Review by mylesdomingo — CHANGES_REQUESTED** [review]

File: src/database/firebase.tsx [hunk 14]
```diff
@@ -8,6 +8,7 @@ import {
   MESSAGING_SENDER_ID,
   APP_ID,
 } from '@env';
+import { Tree, Comment, Additional, Dictionary } from '@types';
```
Comment (wangannie) [inline src/database/firebase.tsx:11]: remove unused import Dictionary

File: src/database/firebase.tsx [hunk 15]
```diff
@@ -18,6 +19,119 @@ const config = {
...
+    return doc.exists;
+  } catch (e) {
+    console.warn(e);
+    throw e;
+    // TODO: Add error handling.
+  }
+};
+
+/*
+ * getTree queries the `trees` table and returns a Tree if the ID is found and an empty entry otherwise.
```
Comment (wangannie) [inline src/database/firebase.tsx:44]: gonna be nitpicky pls address the linter error on the char count!

Comment (wangannie) [reply src/database/firebase.tsx:44]: also for nice comment formatting, we should do 
```
/**
*
*/
```
with double star on first line

**Pull request merged**